assertj-core/src/test/java/org/assertj/core/api/map/MapAssert_isUnmodifiable_Test.java

```diff
@@ -0,0 +1,122 @@
+import static org.assertj.core.api.BDDAssertions.then;
+import static org.assertj.core.error.ShouldBeUnmodifiable.shouldBeUnmodifiable;
+import static org.assertj.core.error.ShouldNotBeNull.shouldNotBeNull;
+import static org.assertj.core.util.AssertionsUtil.expectAssertionError;
+import static org.junit.jupiter.params.provider.Arguments.arguments;
+
+import java.time.temporal.ChronoUnit;
+import java.util.Collections;
+import java.util.EnumMap;
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.stream.Stream;
+
+import org.apache.commons.collections4.map.UnmodifiableMap;
+import org.apache.commons.collections4.map.UnmodifiableSortedMap;
+import org.assertj.core.error.ErrorMessageFactory;
+import org.assertj.core.test.jdk11.Jdk11;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.params.ParameterizedTest;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
+
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSortedMap;
+import com.google.common.collect.Maps;
+
+class MapAssert_isUnmodifiable_Test {
+
+  @Test
+  void should_fail_if_actual_is_null() {
+    // GIVEN
+    Map<?, ?> actual = null;
+    // WHEN
+    AssertionError assertionError = expectAssertionError(() -> assertThat(actual).isUnmodifiable());
+    // THEN
+    then(assertionError).hasMessage(shouldNotBeNull().create());
+  }
+
+  @ParameterizedTest
+  @MethodSource("modifiableMaps")
+  void should_fail_if_actual_can_be_modified(Map<?, ?> actual, ErrorMessageFactory errorMessageFactory) {
+    // WHEN
+    AssertionError assertionError = expectAssertionError(() -> assertThat(actual).isUnmodifiable());
+    // THEN
+    then(assertionError).as(actual.getClass().getName())
+                        .hasMessage(errorMessageFactory.create());
+  }
+
+  private static Stream<Arguments> modifiableMaps() {
+    return Stream.of(arguments(new EnumMap<>(ChronoUnit.class), shouldBeUnmodifiable("Map.clear()")),
+                     arguments(new IdentityHashMap<>(), shouldBeUnmodifiable("Map.clear()")),
+                     arguments(new HashMap<>(), shouldBeUnmodifiable("Map.clear()")),
+                     arguments(new TreeMap<>(), shouldBeUnmodifiable("Map.clear()")),
+                     arguments(new LinkedHashMap<>(), shouldBeUnmodifiable("Map.clear()")));
+  }
+
+  // https://issues.apache.org/jira/browse/COLLECTIONS-xxx FIXME
+  @Test
+  void should_fail_with_commons_collections_UnmodifiableMap() {
+    // GIVEN
+    Map<?, ?> actual = UnmodifiableMap.unmodifiableMap(new HashMap<>());
+    // WHEN
+    AssertionError assertionError = expectAssertionError(() -> assertThat(actual).isUnmodifiable());
+    // THEN
+    then(assertionError).hasMessage(shouldBeUnmodifiable("Map.compute(null, (k, v) -> v)").create());
+  }
```

Comment:
I think this is not a Commons Collections issue but rather caused by the way we test the `compute` method.

The following fails properly:

```java
Map<Object, Object> map = new HashMap<>();
map.put(null, null);
Map<Object, Object> unmodifiableMap = UnmodifiableMap.unmodifiableMap(map);
unmodifiableMap.compute(null, (k, v) -> v); // fails with UOE as it tries to remove (null -> null)
```

but the following doesn't because there is no modification caused to the map:

```java
Map<Object, Object> emptyMap = new HashMap<>();
Map<Object, Object> unmodifiableEmptyMap = UnmodifiableMap.unmodifiableMap(emptyMap);
unmodifiableEmptyMap.compute(null, (k, v) -> v); // does not fail as there is no modification
```

The behavior is also in line with the [`compute`](https://docs.oracle.com/javase/8/docs/api/java/util/Map.html#compute-K-java.util.function.BiFunction-) Javadoc:

> If the remapping function returns null, the mapping is removed (or remains absent if initially absent).

The same is also valid for `unmodifiableSortedMap`.

I'll check how we can improve the assertion for such cases.

Reply:
In retrospect, I don't think we should invest more just to cover two implementations without fail-fast checks. Let's go ahead with the merge and we can always react if concrete use cases are reported.